endpoint/platform/simulator/simulator.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include "common/ta_errors.h"
+#include "endpoint/hal/device.h"
+#include "legato.h"
+
+#include "le_hashmap.h"
+#include "le_log.h"
+
+#define UART_BUFFER_SIZE 1024
+
+/* Setting data to produce predictable results for simulator */
+// device id
+static const char *device_id = "470010171566423";
+// private key
+static const uint8_t private_key[32] = {82,  142, 184, 64,  74, 105, 126, 65,  154, 116, 14,  193, 208, 41,  8,  115,
+                                        158, 252, 228, 160, 79, 5,   167, 185, 13,  159, 135, 113, 49,  209, 58, 68};
+// Buffer for uart
+static char UART_BUFFER[UART_BUFFER_SIZE];
+// Hashmap to simulate secure storage
+static le_hashmap_Ref_t simulator_sec_table;
+
+extern struct device_type simulator_device_type;
+
+static status_t simulator_init(void) {
+  status_t err = register_device(&simulator_device_type);
+  if (err != SC_OK) LE_ERROR("register simulator device error:%d", err);
+  return err;
+}
+
+static void simulator_release(void) {
+  status_t ret = unregister_device(&simulator_device_type);
+  LE_INFO("unregister simulator return: %d", ret);
+}
+
+static status_t simulator_get_key(uint8_t *key) {
+  memcpy(key, private_key, 16);
+  LE_INFO("Get device key success");
+  return SC_OK;
+}
+
+static status_t simulator_get_device_id(char *id) {
+  memcpy(id, device_id, 16);
+  LE_INFO("Get device id success");
+  return SC_OK;
+}
+
+static status_t uart_init(const char *device) {
+  LE_INFO("UART init device %s success", device);
+  return SC_OK;
+}
+
+static void uart_write(const int fd, const char *cmd) {
+  /* simple output */
+  size_t cmd_len = strlen(cmd);
+  if (cmd_len > UART_BUFFER_SIZE) {
+    LE_ERROR("command too long");
+    return;
+  }
+  snprintf(UART_BUFFER, cmd_len, "%s", cmd);
+  LE_INFO("UART write success");
```

Comment:
Since `uart_read` and `uart_write` might be called frequently, there should be no `LE_INFO` for normal usage.